src/core/include/units/reference.h

```diff
@@ -0,0 +1,117 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+#include <units/bits/basic_concepts.h>
+#include <units/bits/dimension_op.h>
+
+namespace units {
+
+template<Dimension D, UnitOf<D> U>
+struct reference;
+
+namespace detail {
+
+template<typename D, typename D1, typename U1, typename D2, typename U2>
+using reference_multiply_impl = reference<D, downcast_unit<D,
+  (U1::ratio / dimension_unit<D1>::ratio) * (U2::ratio / dimension_unit<D2>::ratio) * dimension_unit<D>::ratio>>;
+
+template<typename D, typename D1, typename U1, typename D2, typename U2>
+using reference_divide_impl = reference<D, downcast_unit<D,
+  (U1::ratio / dimension_unit<D1>::ratio) / (U2::ratio / dimension_unit<D2>::ratio) * dimension_unit<D>::ratio>>;
+
+}  // namespace detail
+
+template<Reference R1, Reference R2>
+using reference_multiply = detail::reference_multiply_impl<
+  dimension_multiply<typename R1::dimension, typename R2::dimension>,
+  typename R1::dimension, typename R1::unit, typename R2::dimension, typename R2::unit>;
+
+template<Reference R1, Reference R2>
+using reference_divide = detail::reference_divide_impl<
+  dimension_divide<typename R1::dimension, typename R2::dimension>,
+  typename R1::dimension, typename R1::unit, typename R2::dimension, typename R2::unit>;
+
+/**
+ * @brief The type for unit constants
+ *
+ * This type is intended to be used in the unit constants definition:
+ *
+ * @code{.cpp}
+ * namespace references {
+ * 
+ * inline constexpr auto m = reference<dim_length, metre>{};
+ * inline constexpr auto km = reference<dim_length, kilometre>{};
+ * 
+ * }
+ * @endcode
+ * 
+ * Unit constants simplify quantity creation:
+ * 
+ * @code{.cpp}
+ * using namespace units::isq::si::references;
+ *
+ * auto d = 123 * m;
+ * auto v = 70 * km / h;
+ * @endcode
+ *
+ * Also, it is allowed to define custom unit constants from existing ones:
+ * 
+ * @code{.cpp}
+ * constexpr auto Nm = N * m;
+ * constexpr auto mph = mi / h;
+ * @endcode
+ * 
+ * `km * 3` or `s / 4` syntax is not allowed for quantity creation.
+ */
+template<Dimension D, UnitOf<D> U>
+struct reference {
+  using dimension = D;
+  using unit = U;
+
+#if !UNITS_COMP_MSVC
+  template<typename QuantityOrQuantityValue, typename D2, typename U2>
+    requires Quantity<QuantityOrQuantityValue> || QuantityValue<QuantityOrQuantityValue>
+  friend constexpr Quantity auto operator*(const QuantityOrQuantityValue& lhs, reference<D2, U2>);
+
+  template<typename QuantityOrQuantityValue, typename D2, typename U2>
+    requires Quantity<QuantityOrQuantityValue> || QuantityValue<QuantityOrQuantityValue>
+  friend constexpr Quantity auto operator/(const QuantityOrQuantityValue& lhs, reference<D2, U2>);
+#endif
```

Comment:
Why do we need this? Hidden Friends work only if a definition is provided at this place as well.

Reply:
To make the interface of `reference` clear. Otherwise, it's buried between the operators of `quantity`. It's redundant, though. `quantity_cast` uses `quantity` before it's defined. I didn't do this here as there wasn't a forward declaration of `quantity`. But I should make them hidden friends by forward declaring `quantity`.

Reply:
Well, if a function takes class `A` and class `B` as its parameters then it is hard to determine where it should be defined and should not be defined in two places, right? ;-)

Reply:
Anyway, I believe that `quantity * reference` should not be allowed which means that those operators will disappear from `quantity header` and will land here as hidden friends.